### PR TITLE
 added security-related functionality

### PR DESCRIPTION
**Pull request by nshtykh**

### CODE REVIEW

File: src/main/resources/db/changelog/changes/03-create-roles-table.yaml [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+                  constraints:
+                    primaryKey: true
+                    nullable: false
+              - column:
+                  name: name
+                  type: varchar(255)
+                  constraints:
+                    nullable: false
+                    unique: true
+
```
Comment (MinoDentori) [inline src/main/resources/db/changelog/changes/03-create-roles-table.yaml]: I guess you missed an empty line at the end of the file
Screenshot 2023-12-25 at 09 34 14

**Review by MinoDentori — APPROVED** [review]
Good job!

File: src/main/java/com/example/onlinebookstore/config/SecurityConfig.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    }
+
+    @Bean
+    public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Exception {
+        return http
+                .cors(AbstractHttpConfigurer::disable)
+                .csrf(AbstractHttpConfigurer::disable)
+                .authorizeHttpRequests(
+                        auth -> auth
+                                .requestMatchers("/auth/**", "/error", "/swagger-ui/**")
```
Comment (d1sam) [inline src/main/java/com/example/onlinebookstore/config/SecurityConfig.java]:
```suggestion
                                .requestMatchers("/auth/**", "/error", "/swagger-ui/**", "/v3/api-docs/**")
```

File: src/main/java/com/example/onlinebookstore/controller/BookController.java [hunk 3]
```diff
@@ -27,31 +28,36 @@
 public class BookController {
     private final BookService bookService;
 
+    @PreAuthorize("hasRole('ROLE_USER') or hasRole('ROLE_ADMIN')")
```
Comment (d1sam) [inline src/main/java/com/example/onlinebookstore/controller/BookController.java]:
```suggestion
    @PreAuthorize("hasAnyAuthority('ROLE_USER', 'ROLE_ADMIN')")
```

File: src/main/java/com/example/onlinebookstore/model/User.java [hunk 4]
```diff
@@ -2,16 +2,24 @@
...
 import jakarta.persistence.Table;
+import java.util.Collection;
+import java.util.Set;
 import lombok.Data;
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.core.userdetails.UserDetails;
 
 @Entity
 @Data
 @Table(name = "users")
```
Comment (d1sam) [inline src/main/java/com/example/onlinebookstore/model/User.java:23]:
```suggestion
@Table(name = "users")
@SQLDelete(sql = "UPDATE users SET deleted = true WHERE id=?")
@Where(clause = "deleted=false")
```

File: src/main/java/com/example/onlinebookstore/model/User.java [hunk 5]
```diff
@@ -24,4 +32,44 @@ public class User {
     @Column(nullable = false)
     private String lastName;
     private String shippingAddress;
+    @ManyToMany(fetch = FetchType.EAGER)
+    @JoinTable(name = "users_roles",
+            joinColumns = @JoinColumn(name = "user_id"),
+            inverseJoinColumns = @JoinColumn(name = "role_id"))
+    private Set<Role> roles;
+
```
Comment (d1sam) [inline src/main/java/com/example/onlinebookstore/model/User.java:46]:
```suggestion

  @Column(nullable = false)
  private boolean deleted = false;

```

File: src/main/java/com/example/onlinebookstore/model/User.java [hunk 6]
```diff
@@ -24,4 +32,44 @@ public class User {
...
+    }
+
+    @Override
+    public String getPassword() {
+        return password;
+    }
+
+    @Override
+    public boolean isAccountNonExpired() {
+        return true;
```
Comment (d1sam) [inline src/main/java/com/example/onlinebookstore/model/User.java]:
```suggestion
        return !isDeleted;
```

File: src/main/java/com/example/onlinebookstore/model/User.java [hunk 7]
```diff
@@ -24,4 +32,44 @@ public class User {
...
+    }
+
+    @Override
+    public boolean isAccountNonLocked() {
+        return true;
```
Comment (d1sam) [inline src/main/java/com/example/onlinebookstore/model/User.java]:
```suggestion
        return !isDeleted;
```

File: src/main/java/com/example/onlinebookstore/model/User.java [hunk 8]
```diff
@@ -24,4 +32,44 @@ public class User {
...
+    }
+
+    @Override
+    public boolean isCredentialsNonExpired() {
+        return true;
```
Comment (d1sam) [inline src/main/java/com/example/onlinebookstore/model/User.java]:
```suggestion
        return !isDeleted;
```

File: src/main/java/com/example/onlinebookstore/model/User.java [hunk 9]
```diff
@@ -24,4 +32,44 @@ public class User {
...
+    }
+
+    @Override
+    public boolean isEnabled() {
+        return true;
```
Comment (d1sam) [inline src/main/java/com/example/onlinebookstore/model/User.java]:
```suggestion
        return !isDeleted;
```

File: src/main/java/com/example/onlinebookstore/repository/RoleRepository.java [hunk 10]
```diff
@@ -0,0 +1,7 @@
+package com.example.onlinebookstore.repository;
+
+import com.example.onlinebookstore.model.Role;
+import org.springframework.data.jpa.repository.JpaRepository;
+
```
Comment (d1sam) [inline src/main/java/com/example/onlinebookstore/repository/RoleRepository.java:6]:
```suggestion
@Repository
```

File: src/main/resources/db/changelog/changes/03-create-roles-table.yaml (hunk 1 above)
Comment (d1sam) [reply src/main/resources/db/changelog/changes/03-create-roles-table.yaml]: agree

File: src/main/resources/db/changelog/changes/04-create-users_roles-table.yaml
Comment (d1sam) [inline src/main/resources/db/changelog/changes/04-create-users_roles-table.yaml:1]: Be sure you have at least one user in DB with role admin

**Review by d1sam — CHANGES_REQUESTED** [review]
Good job, but some changes should be done according to my comment

File: src/main/java/com/example/onlinebookstore/model/User.java [hunk 11]
```diff
@@ -37,6 +41,8 @@ public class User implements UserDetails {
             joinColumns = @JoinColumn(name = "user_id"),
             inverseJoinColumns = @JoinColumn(name = "role_id"))
     private Set<Role> roles;
+    @Column(nullable = false)
```
Comment (yevheniisynytsia) [inline src/main/java/com/example/onlinebookstore/model/User.java]: specify the name of the column
in `@Where` you have `is_deleted`, but here `isDeleted`

**Review by yevheniisynytsia — CHANGES_REQUESTED** [review]
well done! one more comments to go

File: src/main/resources/db/changelog/changes/08-insert-data-to-users_roles-table.yaml
Comment (d1sam) [inline src/main/resources/db/changelog/changes/08-insert-data-to-users_roles-table.yaml:1]: try to improve your code and use valueComputed with preConditions for your insert.
Example:
```
      preConditions:
        - sqlCheck:
            expectedResult: 1
            sql: SELECT COUNT(*) FROM users WHERE email='admin@example.com'
        - sqlCheck:
            expectedResult: 1
            sql: SELECT COUNT(*) FROM roles WHERE name='ROLE_ADMIN'
      changes:
        - insert:
            tableName: user_role
            columns:
              - column:
                  name: user_id
                  valueComputed: (SELECT id FROM users WHERE email='admin@example.com')
              - column:
                  name: role_id
                  valueComputed: (SELECT id FROM roles WHERE name='ROLE_ADMIN')
```

**Review by d1sam — CHANGES_REQUESTED** [review]
good job, but some changes in your migration files should be done.

**Review by vovasalyha — APPROVED** [review]

**Pull request merged**